Implement test structure and first preliminary tests

alankbi opened the pull request:
### List of Changes:

* Add integration (API) test and unit test files for existing codebase
* Implement several test cases for SetupService, RunnerService, and other classes
* Create files and basic environment setup for classes that need future test coverage
* Add try finally block to clean up temp folder even if exception occurs

### Notes:

* As of now, the API and generator tests are left blank as I think we need to wait a bit for those to be fleshed out
* RunnerService, SetupService, and some failure tests are implemented (mostly) up to standard (my standards at least lol)
* DockerService is also left blank because it's hard to test (although should be well covered by integration tests by default)

alankbi commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create test structure and preliminary tests for tester service](https://trello.com/c/pPxQDPg9/247-create-test-structure-and-preliminary-tests-for-tester-service)

zpChris commented:
> some failure tests are implemented (mostly) up to standard (my standards at least lol)

Lots of qualifiers there haha 😄

zpChris commented on src/test/java/com/rocketden/tester/api/RunnerTests.java:
```diff
@@ -0,0 +1,59 @@
+        request.setLanguage(LANGUAGE);
+        request.setProblem(new Problem());
+
+        String jsonRequest = "{\"code\": \"print('hi')\", \"language\": \"js\", \"problem\": \"{}\"}";
```
I believe this test fails because there is no language called "js", right -- it's actually called "javascript"? If so, I feel like it'd be better to make this clearer by just making a fake language or a clear typo like "notjava".

zpChris commented on src/test/java/com/rocketden/tester/api/RunnerTests.java:
```diff
@@ -0,0 +1,59 @@
+
+    private static final String POST_RUNNER = "/api/v1/runner";
+
+    private static final String CODE = "public class Solution {}";
+    private static final Language LANGUAGE = Language.JAVA;
```
Remove these fields, as these aren't necessary for the test (as it's currently set up).

zpChris commented on src/test/java/com/rocketden/tester/service/RunnerServiceTests.java:
```diff
@@ -0,0 +1,64 @@
+
+        assertEquals(ERROR, exception.getError());
+        // Despite the exception, the code cleanup step still occurs
+        verify(setupService).deleteTempFolder(TEMP_FOLDER);
```
This was not invoked on when I ran this test locally -- does this test pass for you?

alankbi replied:
Fixed - I forgot to implement a try-finally block to make the cleanup occur regardless of error (I think you mentioned it in an earlier PR)

zpChris commented on src/test/java/com/rocketden/tester/service/generators/JavaDriverGeneratorServiceTests.java:
```diff
@@ -0,0 +1,20 @@
+package com.rocketden.tester.service.generators;
```
Note: I implement tests in this file, and `PythonDriverGeneratorServiceTests.java`, in #9 -- so I have no preference if you keep or delete these files (either way works great).

alankbi replied:
Sounds good - it'll probably get resolved after merging it all

zpChris commented on src/test/java/com/rocketden/tester/util/UtilityTestMethods.java:
```diff
@@ -0,0 +1,16 @@
+        Gson gson = new Gson();
+        return gson.fromJson(json, c);
+    }
+}
```
nit: Add ending newline.

zpChris commented on tester.iml:
```diff
@@ -87,5 +87,7 @@
     <orderEntry type="library" scope="TEST" name="Maven: org.springframework:spring-test:5.2.10.RELEASE" level="project" />
     <orderEntry type="library" scope="TEST" name="Maven: org.xmlunit:xmlunit-core:2.7.0" level="project" />
     <orderEntry type="library" name="Maven: commons-io:commons-io:2.6" level="project" />
+    <orderEntry type="library" name="Maven: commons-lang:commons-lang:2.6" level="project" />
+    <orderEntry type="library" scope="TEST" name="Maven: com.google.code.gson:gson:2.8.6" level="project" />
```
I don't remember having to add these in the main repository -- is this just bad memory or is this for some reason only necessary in `tester.iml`?

alankbi replied:
Yeah no this is an auto-generated file so this isn't relevant to any functioning code stuff

zpChris reviewed (changes requested):
Requesting changes as some of the tests seem to be failing, and I don't think some of these files are necessary. Otherwise, these tests look great though!

alankbi commented on src/test/java/com/rocketden/tester/service/DockerServiceTests.java:
```diff
@@ -0,0 +1,20 @@
+package com.rocketden.tester.service;
```
I'm not mad about skipping these tests at all lol

alankbi commented on src/test/java/com/rocketden/tester/api/JavaTests.java:
```diff
@@ -0,0 +1,63 @@
+    private static final String CODE = "public class Solution {}";
+    private static final Language LANGUAGE = Language.JAVA;
+
+    @Test
+    public void runRequestSuccess() throws Exception {
```
Response in other thread so I'll resolve this one

alankbi commented on src/test/java/com/rocketden/tester/api/PythonTests.java:
```diff
@@ -0,0 +1,63 @@
+	private static final Language LANGUAGE = Language.PYTHON;
+
+	@Test
+	public void runRequestSuccess() throws Exception {
```
Yeah these tests as of now fail - when I wrote them, the driver files were hardcoded and I didn't want to create tests that would only work temporarily. However now that we have the driver code (mostly) working I'll merge the branches and get these ones to pass (although once we figure out the output capturing and flesh out the RunDto return object these tests will have to be updated again)

zpChris replied:
Makes sense and sounds good -- if you wanted to just remove that test or leave it blank until that full process is fully implemented I'd be OK with that, but if you want to update them as we update the codebase that's cool too, your choice 👍

alankbi commented:
Re-requesting review - I pulled the latest changes from `AddProblemDynamic` (also changed this PR to merge into that branch) and since now the driver file stuff mostly works, I implemented several additional tests and error validation checks (so feel free to request changes again cause there's some new stuff)

zpChris commented on src/main/java/com/rocketden/tester/service/SetupService.java:
```diff
@@ -30,16 +30,16 @@ public SetupService(DriverFileService driverFileService) {
     public String createTempFolder(RunRequest request) {
```
Could this function throw an error _after_ it creates the temp folder, but perhaps before it makes the driver or solution file? If so, maybe we could add the `deleteTempFolder` function in the inner try-catch block? (I would comment directly there but Github won't let me.)

alankbi replied:
This is referring to if the createTempFolder method fails halfway through right? If so, I split createTempFolder to add an additional method populateTempFolder, which handles the writing of the driver and solution files. This way, createTempFolder can handle just creating and returning the folder name (which we need in order to be able to call delete on the correct location), and then the populating of file can be included in the try catch block so that the finally block will be executed

zpChris commented on src/main/java/com/rocketden/tester/service/RunnerService.java:
```diff
@@ -18,10 +22,32 @@ public RunnerService(SetupService setupService, DockerService dockerService) {
+        try {
+            return dockerService.spawnAndRun(folder, request.getLanguage());
+        } finally {
+            // Clean up temp folder regardless if above code throws an exception
+            setupService.deleteTempFolder(folder);
```
Thanks for doing this! Good solution ✅

zpChris commented on src/test/java/com/rocketden/tester/api/JavaTests.java:
```diff
@@ -0,0 +1,89 @@
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
```
nit: Remove unused imports.

zpChris commented on src/test/java/com/rocketden/tester/api/JavaTests.java:
```diff
@@ -0,0 +1,89 @@
+package com.rocketden.tester.api;
+
+import com.rocketden.tester.model.problem.ProblemIOType;
+import com.rocketden.tester.model.problem.ProblemInput;
+import com.rocketden.tester.model.problem.ProblemTestCase;
```
nit: Remove unused imports.

zpChris commented on src/test/java/com/rocketden/tester/api/JavaTests.java:
```diff
@@ -0,0 +1,89 @@
+
+    private static final Language LANGUAGE = Language.JAVA;
+    private static final String CODE = String.join("\n",
+            "import java.util.Arrays;",
+            "",
```
This reminds me that we may want to add `import java.util.*;` to the start of every file for the user's code (on frontend and backend) as well.

zpChris commented on src/test/java/com/rocketden/tester/api/JavaTests.java:
```diff
@@ -0,0 +1,89 @@
+            "}");
+
+    @Test
+    public void runRequestSuccess() throws Exception {
```
nit, just an idea: If possible, either in `JavaTests.java` or `PythonTests.java`, could we add tests that verify the Console and Error portions work as well?

I'm not actually sure if this is necessary just because I know this format will change, so I'm 100% ok if we want to go ahead with this test (also because there are 600+ line changes) -- however, I wanted to note it down at least for the future.

alankbi replied:
Yeah, I think it'd def be good to have those (though prob implemented later once we implement the proper output parsing RunDto responses). There's a test stub for error occurred but I'll add ones too for testing console output

zpChris commented on src/test/java/com/rocketden/tester/api/PythonTests.java:
```diff
@@ -0,0 +1,78 @@
+
+	private static final Language LANGUAGE = Language.PYTHON;
+	private static final String CODE = String.join("\n",
+			"class Solution:",
```
This is forward-looking -- similar to Java with `import java.util.*;`, are there any imports we should initially place into the user's file that have commonly-used functions?

alankbi replied:
Good question - I don't use Python for LeetCode-type stuff enough to know if there's anything absolutely critical, but I guess based on that fact I think it should be good? But yeah def important to remember for other languages that might need them

zpChris commented on src/test/java/com/rocketden/tester/service/SetupServiceTests.java:
```diff
@@ -0,0 +1,92 @@
+
+        assertTrue(folder.contains("src/main/java/com/rocketden/tester/temp"));
+
+        // Cleanup any files created (note: this will not always be reached if failures occur)
```
This relates to one of my notes above, but maybe we include `deleteTempFolder` in the exceptions as well to prevent this edge case from happening?

alankbi replied:
I'm not sure if I totally understand the exceptions part, but I did move this code into a try finally block so the deleteTempFolder should now always execute (same thing for the main codebase as we talked about in different comments)

Lmk if you had something else in mind and I can implement that as well (tho maybe in the next PR if this one gets merged before then haha)

zpChris commented on src/test/java/com/rocketden/tester/service/SetupServiceTests.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    @Test
+    public void deleteTempFolderInvalidFolder() {
```
nit: Perhaps test the `DockerSetupError.DELETE_TEMP_FOLDER` error as well?

alankbi replied:
I haven't been able to get that code to really fail based on any of our use cases - it requires us to catch the IOException which is why I added the ApiError, but imo given that it's a production-ready package I feel like we might be able to get away with not testing that

I wasn't overly rigorous on playing around with it tho so if in the future it might be useful then I'd def be down on adding some kind of test for that

zpChris commented on src/test/java/com/rocketden/tester/util/UtilityTestMethods.java:
```diff
@@ -0,0 +1,16 @@
+public class UtilityTestMethods {
+
+    public static String convertObjectToJsonString(Object o) {
+        Gson gson = new Gson();
```
I know we have the same setup in the main repository, and I may have asked this question before, but could we add the GSON object as a static field in this class?

zpChris commented on src/test/java/com/rocketden/tester/util/ProblemTestMethods.java:
```diff
@@ -0,0 +1,43 @@
+public class ProblemTestMethods {
+
+    public static Problem getFindMaxProblem(String... inputs) {
+        Map<Language, String> methodNames = Map.of(
+                Language.JAVA, "findMax",
+                Language.PYTHON, "find_max"
+        );
```
Heads up, per our conversation, you may need to update this if I change the base branch #7.

alankbi replied:
Sounds good, I'll merge main into this branch (it doesn't show any warnings/merge conflicts on GitHub which is concerning given that I immediately get a compilation error locally after merging haha)

zpChris reviewed (approved):
LGTM!

I left some comments and nits, but nothing big, and I'm confident you can make the relevant changes (or decide that they're unnecessary).

zpChris commented:
One more overall nit: There are also some unused imports in `RunnerTests.java`. Do you have a linter that checks for unused imports? No problem if not, just wanted to check ✔️

alankbi commented:
> One more overall nit: There are also some unused imports in `RunnerTests.java`. Do you have a linter that checks for unused imports? No problem if not, just wanted to check ✔️

Removed them - I do have linting haha although I guess I'm prone to miss them at times

Also, there were another few decent rounds of changes made - I'll merge this for now, but if you notice anything feel free to comment and I can address those in future PRs
